File: backend/src/main/java/com/reallink/pump/services/SalesmanBillPaymentService.java
package com.reallink.pump.services;

import java.math.BigDecimal;
import java.util.List;
import java.util.UUID;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.validation.annotation.Validated;

import com.reallink.pump.dto.request.CreateSalesmanBillPaymentRequest;
import com.reallink.pump.dto.request.UpdateSalesmanBillPaymentRequest;
import com.reallink.pump.dto.response.SalesmanBillPaymentResponse;
import com.reallink.pump.entities.Customer;
import com.reallink.pump.entities.PumpInfoMaster;
import com.reallink.pump.entities.SalesmanBillPayment;
import com.reallink.pump.entities.SalesmanShift;
import com.reallink.pump.exception.PumpBusinessException;
import com.reallink.pump.mapper.SalesmanBillPaymentMapper;
import com.reallink.pump.repositories.CustomerRepository;
import com.reallink.pump.repositories.PumpInfoMasterRepository;
import com.reallink.pump.repositories.SalesmanBillPaymentRepository;
import com.reallink.pump.repositories.SalesmanShiftRepository;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import lombok.RequiredArgsConstructor;

@Service
@Validated
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class SalesmanBillPaymentService {

    private final SalesmanBillPaymentRepository repository;
    private final SalesmanShiftRepository salesmanShiftRepository;
    private final CustomerRepository customerRepository;
    private final PumpInfoMasterRepository pumpInfoMasterRepository;
    private final SalesmanBillPaymentMapper mapper;

    public List<SalesmanBillPaymentResponse> getAll() {
        return repository.findAllWithRelations().stream()
                .map(mapper::toResponse)
                .toList();
    }

    public SalesmanBillPaymentResponse getById(@NotNull UUID id) {
        SalesmanBillPayment payment = repository.findByIdWithRelations(id);
        if (payment == null) {
            throw new PumpBusinessException("SALESMAN_BILL_PAYMENT_NOT_FOUND", "Salesman bill payment with ID " + id + " not found");
        }
        return mapper.toResponse(payment);
    }

    @Transactional
    public SalesmanBillPaymentResponse create(@Valid CreateSalesmanBillPaymentRequest request) {
        // Validate pump master exists
        PumpInfoMaster pumpMaster = pumpInfoMasterRepository.findById(request.getPumpMasterId()).orElse(null);
        if (pumpMaster == null) {
            throw new PumpBusinessException("INVALID_PUMP_MASTER",
                    "Pump master with ID " + request.getPumpMasterId() + " does not exist");
        }

        // Validate salesman shift exists and is open
        SalesmanShift salesmanShift = salesmanShiftRepository.findById(request.getSalesmanShiftId()).orElse(null);
        if (salesmanShift == null) {
            throw new PumpBusinessException("INVALID_SALESMAN_SHIFT",
                    "Salesman shift with ID " + request.getSalesmanShiftId() + " does not exist");
        }

        // Validate that the shift is still open (or recently closed for payment recording)
        if (salesmanShift.getStatus() != SalesmanShift.ShiftStatus.OPEN
                && salesmanShift.getStatus() != SalesmanShift.ShiftStatus.CLOSED) {
            throw new PumpBusinessException("INVALID_SHIFT_STATUS",
                    "Cannot record payment for shift with status: " + salesmanShift.getStatus());
        }

        // Validate customer exists
        Customer customer = customerRepository.findById(request.getCustomerId()).orElse(null);
        if (customer == null) {
            throw new PumpBusinessException("INVALID_CUSTOMER",
                    "Customer with ID " + request.getCustomerId() + " does not exist");
        }

        // Create payment entity
        SalesmanBillPayment payment = mapper.toEntity(request);
        payment.setPumpMaster(pumpMaster);
        payment.setSalesmanShift(salesmanShift);
        payment.setCustomer(customer);

        // Save payment
        SalesmanBillPayment savedPayment = repository.save(payment);

        return mapper.toResponse(savedPayment);
    }

    @Transactional
    public SalesmanBillPaymentResponse update(@NotNull UUID id, @Valid UpdateSalesmanBillPaymentRequest request) {
        SalesmanBillPayment existingPayment = repository.findById(id).orElse(null);
        if (existingPayment == null) {
            throw new PumpBusinessException("SALESMAN_BILL_PAYMENT_NOT_FOUND", "Salesman bill payment with ID " + id + " not found");
        }

        // Validate salesman shift if being updated
        if (request.getSalesmanShiftId() != null) {
            SalesmanShift salesmanShift = salesmanShiftRepository.findById(request.getSalesmanShiftId()).orElse(null);
            if (salesmanShift == null) {
                throw new PumpBusinessException("INVALID_SALESMAN_SHIFT",
                        "Salesman shift with ID " + request.getSalesmanShiftId() + " does not exist");
            }

            existingPayment.setSalesmanShift(salesmanShift);
        }

        // Validate customer if being updated
        if (request.getCustomerId() != null) {
            Customer customer = customerRepository.findById(request.getCustomerId()).orElse(null);
            if (customer == null) {
                throw new PumpBusinessException("INVALID_CUSTOMER",
                        "Customer with ID " + request.getCustomerId() + " does not exist");
            }
            existingPayment.setCustomer(customer);
        }

        // Update payment fields using mapper
        mapper.updateEntityFromRequest(request, existingPayment);

        SalesmanBillPayment updatedPayment = repository.save(existingPayment);
        return mapper.toResponse(updatedPayment);
    }

    @Transactional
    public void delete(@NotNull UUID id) {
        if (!repository.existsById(id)) {
            throw new PumpBusinessException("SALESMAN_BILL_PAYMENT_NOT_FOUND", "Salesman bill payment with ID " + id + " not found");
        }
        repository.deleteById(id);
    }

    public List<SalesmanBillPaymentResponse> getBySalesmanShiftId(@NotNull UUID salesmanShiftId) {
        return repository.findBySalesmanShiftId(salesmanShiftId).stream()
                .map(mapper::toResponse)
                .toList();
    }

    public List<SalesmanBillPaymentResponse> getByCustomerId(@NotNull UUID customerId) {
        return repository.findByCustomer_Id(customerId).stream()
                .map(mapper::toResponse)
                .toList();
    }

    public BigDecimal getTotalPaymentsForShift(@NotNull UUID shiftId) {
        return repository.getTotalPaymentsForShift(shiftId);
    }
}

File: frontend/src/services/salesman-bill-payment-service.ts
import api from "./api";
import type {
  CreateSalesmanBillPaymentRequest,
  UpdateSalesmanBillPaymentRequest,
  SalesmanBillPaymentResponse,
} from "@/types";

export class SalesmanBillPaymentService {
  private static readonly BASE_PATH = "/api/v1/salesman-bill-payments";

  // Get all salesman bill payments
  static async getAll(): Promise<SalesmanBillPaymentResponse[]> {
    const response = await api.get<SalesmanBillPaymentResponse[]>(
      this.BASE_PATH
    );
    return response.data;
  }

  // Get payment by ID
  static async getById(id: string): Promise<SalesmanBillPaymentResponse> {
    const response = await api.get<SalesmanBillPaymentResponse>(
      `${this.BASE_PATH}/${id}`
    );
    return response.data;
  }

  // Get payments by salesman bill ID
  static async getByBillId(
    salesmanBillId: string
  ): Promise<SalesmanBillPaymentResponse[]> {
    const response = await api.get<SalesmanBillPaymentResponse[]>(
      `${this.BASE_PATH}/bill/${salesmanBillId}`
    );
    return response.data;
  }

  // Get payments by shift ID
  static async getByShiftId(
    salesmanShiftId: string
  ): Promise<SalesmanBillPaymentResponse[]> {
    const response = await api.get<SalesmanBillPaymentResponse[]>(
      `${this.BASE_PATH}/shift/${salesmanShiftId}`
    );
    return response.data;
  }

  // Get payments by customer ID
  static async getByCustomerId(
    customerId: string,
    pumpMasterId?: string,
    limit?: number
  ): Promise<SalesmanBillPaymentResponse[]> {
    const params = new URLSearchParams();
    if (pumpMasterId) params.append("pumpMasterId", pumpMasterId);
    if (limit) params.append("limit", limit.toString());
    const queryString = params.toString();
    const url = queryString
      ? `${this.BASE_PATH}/customer/${customerId}?${queryString}`
      : `${this.BASE_PATH}/customer/${customerId}`;
    const response = await api.get<SalesmanBillPaymentResponse[]>(url);
    return response.data;
  }

  // Get payments by date range
  static async getByDateRange(
    fromDate: string,
    toDate: string
  ): Promise<SalesmanBillPaymentResponse[]> {
    const response = await api.get<SalesmanBillPaymentResponse[]>(
      `${this.BASE_PATH}/date-range?fromDate=${fromDate}&toDate=${toDate}`
    );
    return response.data;
  }

  // Get total payments for a bill
  static async getTotalByBillId(salesmanBillId: string): Promise<number> {
    const response = await api.get<number>(
      `${this.BASE_PATH}/bill/${salesmanBillId}/total`
    );
    return response.data;
  }

  // Get total payments for a shift
  static async getTotalByShiftId(salesmanShiftId: string): Promise<number> {
    const response = await api.get<number>(
      `${this.BASE_PATH}/shift/${salesmanShiftId}/total`
    );
    return response.data;
  }

  // Create new salesman bill payment
  static async create(
    payment: CreateSalesmanBillPaymentRequest
  ): Promise<SalesmanBillPaymentResponse> {
    const response = await api.post<SalesmanBillPaymentResponse>(
      this.BASE_PATH,
      payment
    );
    return response.data;
  }

  // Update existing salesman bill payment
  static async update(
    id: string,
    payment: UpdateSalesmanBillPaymentRequest
  ): Promise<SalesmanBillPaymentResponse> {
    const response = await api.put<SalesmanBillPaymentResponse>(
      `${this.BASE_PATH}/${id}`,
      payment
    );
    return response.data;
  }

  // Delete salesman bill payment
  static async delete(id: string): Promise<void> {
    await api.delete(`${this.BASE_PATH}/${id}`);
  }
}
